sound/soc/intel/boards/sof_maxim_common.c

```diff
@@ -105,213 +118,163 @@ int max_98373_trigger(struct snd_pcm_substream *substream, int cmd)
 		if (ret)
 			dev_err(rtd->dev,
 				"unable to add Tweeter Left/Right Speaker dapm, ret %d\n", ret);
 	}
-	return ret;
-}
-EXPORT_SYMBOL_NS(max_98390_spk_codec_init, SND_SOC_INTEL_SOF_MAXIM_COMMON);
 
-const struct snd_soc_ops max_98390_ops = {
-	.hw_params = max_98390_hw_params,
+	return ret;
 };
-EXPORT_SYMBOL_NS(max_98390_ops, SND_SOC_INTEL_SOF_MAXIM_COMMON);
+EXPORT_SYMBOL_NS(maxim_spk_codec_init, SND_SOC_INTEL_SOF_MAXIM_COMMON);
 
-void max_98390_set_codec_conf(struct snd_soc_card *card, int ch)
+static int maxim_compute_codec_conf(void)
 {
-	if (ch == ARRAY_SIZE(max_98390_4spk_codec_conf)) {
-		card->codec_conf = max_98390_4spk_codec_conf;
-		card->num_configs = ARRAY_SIZE(max_98390_4spk_codec_conf);
-	} else {
-		card->codec_conf = max_98390_codec_conf;
-		card->num_configs = ARRAY_SIZE(max_98390_codec_conf);
+	const char * const uid_strings[] = { "0", "1", "2", "3" };
+	const char * const MAXIM_HID[] = {"MX98373", "MX98390"};
+	const char * const CODEC_DAI[] = {"max98373", "max98390"};
+
+	unsigned int uid, amp_nums = 0;
+	unsigned int hid, cur = 0;
+	struct acpi_device *adev;
+	struct device *physdev;
+
+	// finding the first match codec HID in MAXIM_HID table.
+	for (uid = 0; uid < MAXIM_MAX_AMPS; uid++) {
+		for(hid = cur; hid < MAXIM_HID_NUMS; hid++) {
+			adev = acpi_dev_get_first_match_dev(MAXIM_HID[hid], uid_strings[uid], -1);
+
+			if (!adev)
+				continue;
+
+			// found, checking matched codec numbers.
+			cur = hid;
+			break;
+		}
+
+		if (!adev) return amp_nums;
+
+		physdev = get_device(acpi_get_first_physical_node(adev));
+
+		if (!physdev)
+			return amp_nums;
+
+		// component info
+		maxim_components[amp_nums].name = dev_name(physdev);
+
+		snprintf(maxim_codec_name, sizeof(maxim_codec_name),
+						"%s-aif1", CODEC_DAI[cur]);
+		maxim_components[amp_nums].dai_name = maxim_codec_name;
+
+		// codec info
+		maxim_codec_conf[amp_nums].dlc.name = dev_name(physdev);
+		maxim_codec_conf[amp_nums].name_prefix = maxim_name_prefixes[uid];
+
+		acpi_dev_put(adev);
+		amp_nums++;
 	}
+
+	if (amp_nums != 2 && amp_nums != 4)
+		pr_warn("Invalid number of amps found: %d, expected 2 or 4\n", amp_nums);
+		return 0;
+
+	return amp_nums;
 }
-EXPORT_SYMBOL_NS(max_98390_set_codec_conf, SND_SOC_INTEL_SOF_MAXIM_COMMON);
+
+/*
+ * 2/4 Maxim codecs
+ */
+void maxim_set_codec_conf(struct snd_soc_card *card)
+{
+	card->codec_conf = maxim_codec_conf;
+	card->num_configs = ARRAY_SIZE(maxim_codec_conf);
+};
+EXPORT_SYMBOL_NS(maxim_set_codec_conf, SND_SOC_INTEL_SOF_MAXIM_COMMON);
+
+void maxim_dai_link(struct snd_soc_dai_link *link)
+{
+	num_codecs = maxim_compute_codec_conf();
+	link->codecs = maxim_components;
+	link->num_codecs = num_codecs;
+	link->init = maxim_spk_codec_init;
+	link->ops = &maxim_ops;
+};
+EXPORT_SYMBOL_NS(maxim_dai_link, SND_SOC_INTEL_SOF_MAXIM_COMMON);
+
+const struct snd_soc_ops maxim_ops = {
+	.hw_params = maxim_hw_params,
+	.trigger = maxim_trigger,
+};
+EXPORT_SYMBOL_NS(maxim_ops, SND_SOC_INTEL_SOF_MAXIM_COMMON);
```

Comment:
i give up, I don't know how you organized the changes but it's just not easy to review. Please try to organize things in smaller patches where we can see what changes you are making. This sort of big code reshuffling is error-prone and impossible to review.

Reply:
@plbossart, before I submit desired smaller patches, share my idea first.
Maybe I should separate this function relevance out as one commit for easier review?

The maxim_compute_codec_conf handles min 2 codecs(L/R, default ) and max 4 codecs(R/L/TR/TL).

Iti is scanning each on board codec/uid if found in MAXIM_HID[]/uid_strings[] and assigning corresponding dai_name from CODEC_DAI[].

for example:
2 codecs available cases:
dev_name, prefix, dai_name
ADS8396-00 - > "Right" , (max98396-aif1)
ADS8396-01 -> "Left"    , (max98396-aif1)

4 codecs available cases:
max98390-00 - > "Right", (max98390-aif1)
max98390-01 -> "Left"    , (max98390-aif1)
max98390-02 -> "Tweeter Right"  ,  (max98390-aif1)
max98390-03 -> "Tweeter Left"    ,  (max98390-aif1)